ext/mysql2/extconf.rb

```diff
@@ -1,188 +1,312 @@
 # encoding: UTF-8
 require 'mkmf'
 
-def asplode lib
-  if RUBY_PLATFORM =~ /mingw|mswin/
-    abort "-----\n#{lib} is missing. please check your installation of mysql and try again.\n-----"
-  elsif RUBY_PLATFORM =~ /darwin/
-    abort "-----\n#{lib} is missing. Try 'brew install mysql', check your installation of mysql and try again.\n-----"
-  else
-    abort "-----\n#{lib} is missing. Try 'apt-get install libmysqlclient-dev' or
-'yum install mysql-devel', check your installation of mysql and try again.\n-----"
-  end
+class << self
+   alias_method :rm_f_original, :rm_f
+   require 'rake'
+   alias_method :rm_f, :rm_f_original
 end
 
-# 2.0-only
-have_header('ruby/thread.h') && have_func('rb_thread_call_without_gvl', 'ruby/thread.h')
+class Platform
+  # Gets the proper platform object for the current platform.
+  def self.current
+    case RUBY_PLATFORM
+      when /mswin/
+        Windows.new
+      when /mingw/
+        WindowsMingw.new
+      when /darwin/
+        MacOS.new
+      else
+        Linux.new
+    end
+  end
 
-# 1.9-only
-have_func('rb_thread_blocking_region')
-have_func('rb_wait_for_single_fd')
-have_func('rb_hash_dup')
-have_func('rb_intern3')
+  def configure
+    _, @rpath_dir = detect_paths
+    detect_headers
+    detect_libraries
+
+    configure_compiler
+  end
+
+  protected
+
+  def rpath_dir
+    @rpath_dir
+  end
+
+  # @return [(String, String)|nil] The include and library paths.
+  def detect_paths
+    detect_by_explicit_path || detect_by_mysql_config
+  end
+
+  def detect_headers
+    headers = %w{ errmsg.h mysqld_error.h }
+    prefix = ['', 'mysql/'].find do |prefix|
+      have_header("#{prefix}mysql.h")
+    end
+    asplode('mysql.h') unless prefix
+
+    headers.each do |header|
+      header = "#{prefix}#{header}"
+      asplode(header) unless have_header(header)
+    end
+  end
+
+  def detect_libraries
+    libraries = %w{ mysqlclient libmysql }
+    library = libraries.find do |library|
+      have_library(library, 'mysql_query')
```

Comment:
This method only handles the detection of libraries if there is no mysql-config path specified.